### ISSUE TITLE
Capturing Asset Thumbnail and Exporting to bmp

### ISSUE THREAD

**Issue by RockShea** [20tab/UnrealEnginePython#492]
I am attempting to capture the thumbnail of a given skeletal mesh and export it as an image format for use in an asset importer tool in multiple DCC's.
I have gotten as far as generating the thumbnail and then saving it as it's own asset. 
When I go to export it from the right click menu it only gives me two text based options.

```
import os
import unreal_engine as ue
from unreal_engine.classes import TextureExporterBMP, Texture

def get_FileName(inFileName, includeExtension=True):
	"""
		Get the name of the given file.
	"""

	if includeExtension:
		
		return os.path.basename(inFileName)

	else:

		return os.path.splitext(os.path.basename(inFileName))[0]
		
def get_ParentDirectory(inPath, returnFullPath=True, includeExtension=True):
	"""
		Get the parent directory of the given path.
	"""

	if returnFullPath:
		
		return os.path.abspath(os.path.join(inPath, os.pardir))

	else:

		return get_FileName(os.path.abspath(os.path.join(inPath, os.pardir)))

selectedAssets = ue.get_selected_assets()

def get_AssetPath(inAsset, includeTrailingBackslash=True):

	assetPath = asset.get_path_name()
	assetPath = assetPath.split(".")[0]

	assetPath = get_ParentDirectory(assetPath).replace("C:", "")
	if includeTrailingBackslash:
		
		assetPath = assetPath + "\\"
	
	return assetPath

def make_UnrealPath(inPath):
	
	return inPath.replace("\\", "/")

if __name__ == '__main__':

	for asset in selectedAssets:
		
		assetPath      = get_AssetPath(asset)
		print(assetPath)
		assetName      = asset.get_name()
		assetThumb     = asset.get_thumbnail()
		assetThumbData = assetThumb.get_uncompressed_image_data()
		texture        = ue.create_transient_texture(assetThumb.get_image_width(), assetThumb.get_image_height())
		texture.texture_set_data(assetThumbData)
		texturePathUnreal = make_UnrealPath(assetPath + assetName + "Thumb")
		print(texturePathUnreal)
		print(texture.save_package(texturePathUnreal))
		texture.set_name(assetName + "Thumb")
```

Is there something I am missing or is this functionality not expected to fully work just yet?

**Comment by rdeioris** [comment]
Hi, transient textures are, well, transient :) It means you cannot save them on disk and, more important their data are only available in the GPU. You need to use the create_texture() function to build the asset and texture_set_source_data() to assign the rgba values.

**Comment by RockShea** [comment]
Thanks for the reply!
I had taken snippets from the thumbnail example file and given no thought to the transient keyword in the method name, smart I know.

Having swapped out the that method call for the correct one; create_texture, it requires 5 arguments and I am unable to figure out what they may be, I have peeked into the source code to try and figure it out from there but am having no luck :( https://github.com/20tab/UnrealEnginePython/commit/a0564031e88dc4a974124db3b27832e99df1d0d2
The first argument I would have thought would be a string to the package path, but I get an error:
```
argument is not a UPackage
Traceback (most recent call last):
  File "<string>", line 33, in <module>
Exception: argument is not a UPackage
```

Sorry, it must be painful hand holding to such a degree.

**Comment by rdeioris** [comment]
Hi, you can leave the package argument as  None:

```python
tex = ue.create_texture(None, 'Foo', width, height)
```

**Comment by RockShea** [comment]
Hey, thanks again for the reply!

Here is the updated code which saves the selected objects thumb as a texture in the same location.
It seems to work, however the texture it saves is black.
My thoughts are it is something to do with the 5th argument of the function which requires a bytes-like object if not set as a string or None:

`
import os
import unreal_engine as ue
from unreal_engine.classes import TextureExporterBMP, Texture
from R_Shared.Core import R_Path
import inspect


def get_FileName(inFileName, includeExtension=True):
	"""
	Get the name of the given file.
	"""

	if includeExtension:
		
		return os.path.basename(inFileName)

	else:

		return os.path.splitext(os.path.basename(inFileName))[0]
		
def get_ParentDirectory(inPath, returnFullPath=True, includeExtension=True):
	"""
	Get the parent directory of the given path.
	"""

	if returnFullPath:
		
		return os.path.abspath(os.path.join(inPath, os.pardir))

	else:

		return get_FileName(os.path.abspath(os.path.join(inPath, os.pardir)))

def get_AssetPath(inAsset, includeTrailingBackslash=True):

	assetPath = inAsset.get_path_name()
	assetPath = assetPath.split(".")[0]

	assetPath = get_ParentDirectory(assetPath).replace("C:", "")
	if includeTrailingBackslash:
		
		assetPath = assetPath + "\\"
	
	return assetPath

def make_UnrealPath(inPath):
	
	return inPath.replace("\\", "/")

def get_ThumbAsTexture():

	selectedAssets = ue.get_selected_assets()

	for asset in selectedAssets:
		
		assetPath         = get_AssetPath(asset)
		assetName         = asset.get_name()
		assetThumb        = asset.get_thumbnail()
		assetThumbData    = assetThumb.get_uncompressed_image_data()
		texture           = ue.create_texture(None, assetName + "Thumb", assetThumb.get_image_width(), assetThumb.get_image_height(), None)
		texturePathUnreal = make_UnrealPath(assetPath + assetName + "Thumb")
		texture.texture_set_data(assetThumbData)
		texture.texture_set_data(assetThumbData)
		texture.save_package(texturePathUnreal)

if __name__ == '__main__':

	get_ThumbAsTexture()
`

**Comment by rdeioris** [comment]
You for non-transient texture you need to use texture_set_source_data() to assign bytes. The fifth argument of creat_texture is the bytestream itself (so eventually you can avoid to call texture_set_source_data())

**Comment by RockShea** [comment]
Of course, that makes perfect sense, I will give this a go once my build finishes.
Thanks again!

**Comment by RockShea** [comment]
Thanks very much for all the help.
Here is the final working script:

```
import os
import unreal_engine as ue
from unreal_engine.classes import TextureExporterBMP

textureExporterBMP = TextureExporterBMP()

def get_AssetPath(inAsset, includeTrailingBackslash=True):

	assetPath = inAsset.get_path_name()
	assetPath = assetPath.split(".")[0]

	assetPath = os.path.abspath(os.path.join(assetPath, os.pardir)).replace("C:", "")
	if includeTrailingBackslash:
		
		assetPath = assetPath + "\\"
	
	return assetPath

def make_UnrealPath(inPath):
	
	return inPath.replace("\\", "/")

def export_ThumbAsBmp(inAssets, inSavePath):
	
	if not inAssets:
	
		inAssets = ue.get_selected_assets()

	if not isinstance(inAssets, (list, tuple)):
	
		inAssets = [inAssets]

	for asset in inAssets:
		
		assetPath         = get_AssetPath(asset)
		assetName         = asset.get_name()
		assetThumb        = asset.get_thumbnail()
		assetThumbData    = assetThumb.get_uncompressed_image_data()
		texture           = ue.create_texture(None, assetName + "Thumb", assetThumb.get_image_width(), assetThumb.get_image_height(), assetThumbData)
		textureExporterBMP.export_to_file(texture, os.path.abspath(os.path.join(inSavePath, assetName + ".bmp")))

if __name__ == '__main__':

	export_ThumbAsBmp(None, r"C:\Data")

```
